ansible/roles/dev-desktop/files/git-credential-dev-desktop-inner

```diff
@@ -0,0 +1,146 @@
+#!/usr/bin/python3 -I
+# The `-I` flag enables isolation, which disables environment variables that
+# change the behavior of the interpreter like `PYTHONPATH`, preventing
+# privilege escalation (as this script is meant to be executed through sudo).
+
+from sys import argv, stdin, stderr
+import jwt
+import os
+import requests
+import time
+
+
+def generate_jwt():
+    base = os.path.dirname(os.path.realpath(__file__))
+    with open('/etc/github-app-credentials/app_id.txt', 'r') as fh:
+        app_id = int(fh.read())
+    with open('/etc/github-app-credentials/app_private_key.pem', 'rb') as fh:
+        private_key = fh.read()
+
+    now = int(time.time())
+    payload = {
+        "iat": now - 60,
+        "exp": now + (60 * 9),
+        "iss": app_id,
+    }
+
+    token = jwt.encode(payload, private_key, algorithm="RS256")
+    if type(token) == bytes:
+        token = token.decode("ascii")
+    return token
+
+
+def generate_token(user, repo):
+    token = generate_jwt()
+
+    def req(method, url):
+        return requests.request(method, f"https://api.github.com/{url}", headers={
+            "Authorization": f"Bearer {token}",
+        })
+
+    installation = req("GET", f"repos/{user}/{repo}/installation")
+    if installation.status_code == 404:
+        # 404 could be caused either by the app not being installed at all on the account,
+        # or by the repository not being authorized with the app. Try to retrieve the
+        # installation in the user's account to check which case we're handling.
+        user_installation = req("GET", f"users/{user}/installation")
+        if user_installation.status_code == 200:
+            raise RepositoryNotAuthorizedError(user_installation.json()["html_url"])
+        else:
+            # Requesting the app is required to get the app URL.
+            app = req("GET", "app")
+            app.raise_for_status()
+            raise ApplicationNotInstalledError(app.json()["html_url"])
+    installation.raise_for_status()
+    installation = installation.json()
+
+    auth = req("POST", f"app/installations/{installation['id']}/access_tokens")
+    if auth.status_code == 403:
+        raise SuspendedInstallationError(installation["html_url"])
+    auth.raise_for_status()
+    auth = auth.json()
+
+    return auth["token"]
+
+
+def credential_provider():
+    if argv[1] != "get":
+        exit(0)
+
+    config = dict(x.strip().split("=", 1) for x in stdin)
+
+    if "host" not in config or config["host"] != "github.com":
+        exit(0)
+    git_user, repo = config["path"].split("/", 1)
+    if repo.endswith(".git"):
+        repo = repo[:-len(".git")]
+
+    # Get the username of the caller, not the owner of
+    # this script (which is setuid)
+    user = os.getlogin()
```

Comment:
From https://linux.die.net/man/3/getlogin_r (emphasis mine)

> Unfortunately, it is often rather easy to fool getlogin(). Sometimes it does not work at all, because some program messed up the utmp file. Often, it gives only the first 8 characters of the login name. The user currently logged in on the controlling terminal of our program need not be the user who started it. **Avoid getlogin() for security-related purposes.**

Reply:
Rebased on top of master, and added the last commit to address this. I am *pretty sure* that's safe, but I'd like another set of eyes on it.